### PR TITLE
Introduce custom fields in OAuth2-Token-Error-Response

### PR DESCRIPTION
**Pull request by dulanjalidilmi**
### Proposed changes in this pull request
With this change, the user can add custom parameters during an error scenario when using a custom grant handler.

Fix: https://github.com/wso2/api-manager/issues/808

### When should this PR be merged
N/A

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpoint.java [hunk 1]
```diff
@@ -259,11 +259,28 @@ private Response handleErrorResponse(OAuth2AccessTokenRespDTO oauth2AccessTokenR
...
                     .errorResponse(HttpServletResponse.SC_BAD_REQUEST)
                     .setError(oauth2AccessTokenResp.getErrorCode())
-                    .setErrorDescription(oauth2AccessTokenResp.getErrorMsg())
-                    .buildJSONMessage();
+                    .setErrorDescription(oauth2AccessTokenResp.getErrorMsg());
+
+            try {
+                if (MapUtils.isNotEmpty(oauth2AccessTokenResp.getErrorParameterMap())) {
+                    for (Map.Entry<String, Object> entry: oauth2AccessTokenResp.getErrorParameterMap().entrySet()) {
+                        if (entry.getValue() instanceof String) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpoint.java:270]: we are doing the same action in both if and else

Comment (mefarazath) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpoint.java:270]: Do we need a branch?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AccessTokenRespDTO.java [hunk 2]
```diff
@@ -189,6 +191,24 @@ public Map<String, String> getParameters() {
         return Collections.unmodifiableMap(getParameterMap());
     }
 
+    /**
+     * Get error parameter map
+     *
+     * @return a map of error parameters
+     */
+    public Map<String, Object> getErrorParameterMap() {
+        return errorParameterMap;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AccessTokenRespDTO.java:200]: Would be good to return an unmodifiable map and introduce a seperate method to add a new error param

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AccessTokenRespDTO.java [hunk 3]
```diff
@@ -189,6 +191,24 @@ public Map<String, String> getParameters() {
...
+    }
+
+    /**
+     * Set error parameter map
+     *
+     * @param errorParameterMap
+     */
+    public void setErrorParameterMap(Map<String, Object> errorParameterMap) {
+        this.errorParameterMap = errorParameterMap;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AccessTokenRespDTO.java:209]:
```suggestion

        this.errorParameterMap = errorParameterMap;
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpoint.java (hunk 1 above)
Comment (dulanjalidilmi) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpoint.java]: Fixed. 

For APIM 4.2.0 release, we will be using 6.9.x branch.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OAuthTokenReqMessageContext.java [hunk 4]
```diff
@@ -171,4 +175,39 @@ public void setConsentedToken(boolean consentedToken) {
...
+
+        return Collections.unmodifiableMap(getErrorMap());
+    }
+
+    /**
+     * Add key values to error paramter map
+     * @param key key
+     * @param value value
+     */
+    public void setErrorParameterMap(String key, Object value) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OAuthTokenReqMessageContext.java:194]: Shall we revisit the name?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AccessTokenRespDTO.java (hunk 2 above)
Comment (dulanjalidilmi) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AccessTokenRespDTO.java]: Fixed.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AccessTokenRespDTO.java (hunk 3 above)
Comment (dulanjalidilmi) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dto/OAuth2AccessTokenRespDTO.java]: Fixed.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OAuthTokenReqMessageContext.java (hunk 4 above)
Comment (dulanjalidilmi) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/OAuthTokenReqMessageContext.java]: Updated to `addErrorParameter`

File: components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpointTest.java [hunk 5]
```diff
@@ -389,6 +389,85 @@ public void testTokenErrorResponse(String errorCode, Object headerObj, int expec
...
+            response = oAuth2TokenEndpoint.issueAccessToken(request, new MultivaluedHashMap<String, String>());
+        } catch (InvalidRequestParentException ire) {
+            InvalidRequestExceptionMapper invalidRequestExceptionMapper = new InvalidRequestExceptionMapper();
+            response = invalidRequestExceptionMapper.toResponse(ire);
+        }
+
+        assertNotNull(response, "Token response is null");
+        assertEquals(response.getStatus(), expectedStatus, "Unexpected HTTP response status");
+        assertNotNull(response.getEntity(), "Response entity is null");
+        assertTrue(response.getEntity().toString().contains(expectedErrorCode), "Expected error code not found");
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpointTest.java:464]: can we add assert for the header values as well here as we are changing that logic as well

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java [hunk 6]
```diff
@@ -905,15 +906,40 @@ private OAuth2AccessTokenRespDTO handleError(String errorCode,
...
+        tokenRespDTO.setResponseHeaders(headers);
+    }
+
+    /**
+     * Set custom error parameters in OAuth2AccessTokenRespDTO
+     * @param tokenRespDTO
+     * @param errorMap
+     */
+    private void setCustomErrorParameters(OAuth2AccessTokenRespDTO tokenRespDTO, Map<String, Object> errorMap) {
+        for (Map.Entry<String, Object> entry: errorMap.entrySet()) {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:923]: Need a new line after method definition

File: components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpointTest.java [hunk 7]
```diff
@@ -389,6 +389,85 @@ public void testTokenErrorResponse(String errorCode, Object headerObj, int expec
...
+        contentType.setValue(OAuth.ContentType.URL_ENCODED);
+
+        ResponseHeader[] headers = new ResponseHeader[]{contentType};
+
+        Map<String, String> customErrorsMap = new HashMap<>();
+        customErrorsMap.put("customMessage1", "This is custom message 1");
+        customErrorsMap.put("customMessage2", "This is custom message 2");
+
+        return new Object[][]{
+                {TOKEN_ERROR, headers, customErrorsMap, HttpServletResponse.SC_BAD_REQUEST, TOKEN_ERROR}
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpointTest.java:406]: We can add some other combinations with empty error map, empty headers to the data provider

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java (hunk 6 above)
Comment (dulanjalidilmi) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:924]: Fixed.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpointTest.java (hunk 5 above)
Comment (dulanjalidilmi) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpointTest.java:470]: Fixed.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpointTest.java (hunk 7 above)
Comment (dulanjalidilmi) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpointTest.java]: Fixed.

**Review by senthalan — APPROVED** [review]

**Pull request merged**